sike.go

```diff
@@ -0,0 +1,174 @@
+/*
+Abstract:
+This file contains APIs for Supersingular Isogeny Key Encapsulation (SIKE) protocol which is recently submitted to NIST PQC candidates.
+The underlying core functions are based on SIDH keygeneration functions developed by Cloudflare.
+SIKE requires NIST's approved hash functions such as sha3 and cSHAKE256 to encapsulate the key based on NIST specifications. The custome 
+SHAKE256 is adopted from  https://gist.github.com/mimoo/7e815318e54d5c07c3330149ddf439c5
+and placed inside "golang.org/x/crypto/sha3" path
+
+Author: Amir Jalali  				ajalali2016@fau.edu
+Date: Feb 2018
+*/
+
+package p751sidh
+
+import (
+	"io"
+	"golang.org/x/crypto/sha3"
+	"bytes"
+)
+
+const (
+	// The message size, in bytes.
+	MessageSize = 32
+	// The ciphertext size, in bytes = PublicKeySize + MessageSize
+	CiphertextSize = 596
+	// SIKE secret key size, in bytes = MessageSize + PublicKeySize + SecretKeySize
+	SIKESecretKeySize = 644
+)
+
+// SIKE secret key.
+type SIKESecretKey struct {
+	//Scalar Message
+	Scalar [MessageSize]byte
+	//SecretKey Bob
+	SecretKey *SIDHSecretKeyBob
+	//PublicKey Bob
+	PublicKey *SIDHPublicKeyBob
+}
+
+type SIKECipherText struct {
+	//PublicKey Alice
+	PublicKey *SIDHPublicKeyAlice
+	//Scalar Message
+	Scalar [MessageSize]byte
+}
+
+type SIKESharedSecret struct{
+
+	Scalar [SharedSecretSize]byte
+}
+
+func crypto_SIKE_keypair(rand io.Reader) (publicKey *SIDHPublicKeyBob, SIKEsecretKey *SIKESecretKey, err error){
+	// SIKE keypair generation
+	// Outputs: SIKE secret key (message + secretkey + publickey)
+	//			SIKE public key (public key)
+
+	publicKey = new(SIDHPublicKeyBob)
+	SIKEsecretKey = new(SIKESecretKey)
+	var secretKey = new(SIDHSecretKeyBob)
+
+	//first randomly generate 32 byte s
+	_, err = io.ReadFull(rand, SIKEsecretKey.Scalar[:])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	//next, generate Encryptor secret-key and public-key
+	publicKey, secretKey, err = GenerateBobKeypair(rand)
+
+	//Copy secretKey and publicKey into SIKEsecretKey
+	SIKEsecretKey.SecretKey = secretKey
+	SIKEsecretKey.PublicKey = publicKey
+
+	return
+}
+
+func crypto_SIKE_encap(rand io.Reader, publicKey *SIDHPublicKeyBob) (cipherText *SIKECipherText, sharedSecret *SIKESharedSecret, err error) {
+	// SIKE encapsulation
+	// Input: publicKey
+	// Output: shared secret ss
+	//		   ciphertext message ct (publicKey + message)
+	cipherText = new(SIKECipherText)
+	sharedSecret = new(SIKESharedSecret)
+
+	const G = "0"
+	const H = "1"
+	const P = "2"
+	var ephemeral_sk = new(SIDHSecretKeyAlice)
+	var jinvariant [SharedSecretSize]byte
+	var h_ [MessageSize]byte
+	//var tmp [CiphertextSize + MessageSize]byte
+	var tmp = make([]byte, (CiphertextSize + MessageSize))
+
+	//Generate ephemeral secretKey G(m||pk) mod oA
```

Comment:
Please apply gofmt and golint to catch stylistic issues like this (no space after slash in comment).